### PR TITLE
feat: job steps

### PR DESCRIPTION
**Pull request by rephus**
### Pre requirements:
- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/lightdash/lightdash/blob/main/.github/CONTRIBUTING.md#sending-a-pull-request).

### Reference:
Closes: #1918

### Description:
Add steps to Jobs, Migrates job methods into JobModel

### Preview:
DB updates 
Peek 2022-04-28 14-18

new /job steps

Peek 2022-04-28 14-19

### Scope of the changes (select all that apply):
- [ ] Frontend  
- [x] Backend  
- [ ] Common  
- [ ] E2E tests  
- [ ] Docs  
- [ ] CI/CD

### CODE REVIEW

File: packages/backend/src/services/ProjectService/ProjectService.ts
Comment (rephus) [inline packages/backend/src/services/ProjectService/ProjectService.ts:107]: made a new async function outside `create` to make it async

File: packages/backend/src/services/ProjectService/ProjectService.ts
Comment (rephus) [inline packages/backend/src/services/ProjectService/ProjectService.ts:152]: called without await, this should make create mostly async, just returning the jobuuid

File: packages/backend/src/database/migrations/20220428071116_job_steps.ts
Comment (rephus) [inline packages/backend/src/database/migrations/20220428071116_job_steps.ts]: project_uuid needs to be nullable, because we will be calling it now before the project extists

File: packages/backend/src/database/migrations/20220428071116_job_steps.ts
Comment (rephus) [inline packages/backend/src/database/migrations/20220428071116_job_steps.ts:9]: project_uuid needs to be nullable now, because we start job before project exists

File: packages/backend/src/database/migrations/20220428071116_job_steps.ts
Comment (rephus) [inline packages/backend/src/database/migrations/20220428071116_job_steps.ts:7]: there is a better new way to make it nullable now, but it is not available here

File: packages/backend/src/models/ProjectModel/ProjectModel.ts
Comment (rephus) [inline packages/backend/src/models/ProjectModel/ProjectModel.ts:471]: copied from job, it wil be removed

File: packages/backend/src/models/ProjectModel/ProjectModel.ts
Comment (rephus) [inline packages/backend/src/models/ProjectModel/ProjectModel.ts:479]: TODO also update the job here

File: packages/backend/src/models/ProjectModel/ProjectModel.ts
Comment (rephus) [reply packages/backend/src/models/ProjectModel/ProjectModel.ts:469]: rudderstack event names ?

File: packages/backend/src/services/ProjectService/ProjectService.ts
Comment (rephus) [inline packages/backend/src/services/ProjectService/ProjectService.ts:151]: TODO from discussion: Create here full list of steps that we will update later

File: packages/backend/src/services/ProjectService/ProjectService.ts
Comment (rephus) [inline packages/backend/src/services/ProjectService/ProjectService.ts:111]: TODO from discussion: Create here full list of steps that we will update later

**Comment by rephus** [comment]
We can't merge this until Frontend is done. Otherwise the /createProject breaks the onboarding !

**Review by ZeRego — CHANGES_REQUESTED** [review]

File: packages/backend/src/database/entities/jobs.ts
Comment (rephus) [reply packages/backend/src/database/entities/jobs.ts:23]: good catch, I added updated_at recently

File: packages/backend/src/models/JobModel/JobModel.ts
Comment (rephus) [reply packages/backend/src/models/JobModel/JobModel.ts:137]: Yeah, I don't know where this will happen, this is a  remindert o add it at some point, probably on a separate branch

File: packages/backend/src/models/JobModel/JobModel.ts
Comment (rephus) [reply packages/backend/src/models/JobModel/JobModel.ts:109]: can't I just use a regular for loop then ? that feels easier than adding an extra column in db

Comment (ZeRego) [reply packages/backend/src/models/JobModel/JobModel.ts:109]: yeah, that also works 👍

File: packages/backend/src/database/entities/jobs.ts
Comment (owlas) [inline packages/backend/src/database/entities/jobs.ts:22]: We need a column to store the error if the step fails, or is that something that would live under the step label?

File: packages/backend/src/models/JobModel/JobModel.ts
Comment (owlas) [inline packages/backend/src/models/JobModel/JobModel.ts:108]: You can use a single insert with an array of records! That way you don't need to loop at all: https://knexjs.org/#Builder-insert

If there's concern that you want to be able to retrieve them in exactly the right order, SO tells me [postgres doesn't guarantee it'll insert in the same order you specify but it actually does](https://dba.stackexchange.com/questions/95822/does-postgres-preserve-insertion-order-of-records) - we'd have to add an extra column with the order id if we really want to be absolutely certain. But I'm happy to rely on postgres insert order tbh

**Comment by rephus** [comment]
Updated with tryJobStep , nice suggestion :ok_hand:

File: packages/backend/src/models/JobModel/JobModel.ts
Comment (owlas) [inline packages/backend/src/models/JobModel/JobModel.ts:111]: Better not to store the step label if it's simply derived from the step type. Plus it means we can always change the labels shown to the user without affecting the db at all. Win!

File: packages/backend/src/database/migrations/20220428071116_job_steps.ts
Comment (owlas) [inline packages/backend/src/database/migrations/20220428071116_job_steps.ts:12]: there is no constraint here that the jobs can only have one of each `step_type` - would it be needed?

If we can insert the same step twice, then indexing by (`job_uuid`, `step_type`) could return multiple records from job steps

File: packages/backend/src/models/JobModel/JobModel.ts
Comment (owlas) [reply packages/backend/src/models/JobModel/JobModel.ts]: Can we actually derive the whole job status from the steps??

If so maybe we can actually completely ignore this field and just reduce the step status, same for `updated_at` - that way we only every need to update the steps

File: packages/backend/src/models/JobModel/JobModel.ts
Comment (owlas) [inline packages/backend/src/models/JobModel/JobModel.ts:167]: If we store the error message in the job, then the whole thing can complete async, and we'd discover the error by checking the job. wdyt?

@ZeRego @rephus we didn't actually discuss what would happen on error but I assumed that's how it would work

Comment (owlas) [reply packages/backend/src/models/JobModel/JobModel.ts:167]: I think we would only store errors from the `callback`

File: packages/backend/src/routers/organizationRouter.ts
Comment (owlas) [inline packages/backend/src/routers/organizationRouter.ts:58]: so this is currently breaking the UI and we can't update the UI until we merge @nathaliabj 's work?

File: packages/backend/src/services/ProjectService/ProjectService.ts
Comment (owlas) [inline packages/backend/src/services/ProjectService/ProjectService.ts:121]: beautiful, so this kicks off a background job basically. But this is breaking the api.

But I think the approach is great

File: packages/backend/src/services/ProjectService/ProjectService.ts
Comment (owlas) [inline packages/backend/src/services/ProjectService/ProjectService.ts:126]: This reads so nicely! Makes business logic very obvious

**Review by owlas — CHANGES_REQUESTED** [review]

File: packages/backend/src/routers/organizationRouter.ts
Comment (rephus) [reply packages/backend/src/routers/organizationRouter.ts:58]: correct

File: packages/backend/src/database/entities/jobs.ts
Comment (rephus) [reply packages/backend/src/database/entities/jobs.ts:22]: I can rename step_label to `step_error : string | undefined` if you want to store errors but don't want to store labels and prefer to generate on the fly

File: packages/backend/src/services/ProjectService/ProjectService.ts
Comment (ZeRego) [inline packages/backend/src/services/ProjectService/ProjectService.ts:163]: I think we are missing setting the job status to done at the end of all these steps.

File: packages/backend/src/database/migrations/20220428071116_job_steps.ts
Comment (rephus) [reply packages/backend/src/database/migrations/20220428071116_job_steps.ts:12]: image

File: packages/backend/src/database/migrations/20220428071116_job_steps.ts
Comment (owlas) [inline packages/backend/src/database/migrations/20220428071116_job_steps.ts:36]: Hot dang!!!! We got a constraint!

This will also setup an index on this pair so makes lookups much faster 😀

**Review by owlas — CHANGES_REQUESTED** [review]

**Pull request merged**
